### PR TITLE
Security

### PR DESCRIPTION
**Pull request by vfilimo**

### CODE REVIEW

File: src/main/resources/db/changelog/changes/07-insert-users.yaml [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+            columns:
+              - column:
+                  name: id
+                  valueNumeric: 1
+              - column:
+                  name: email
+                  value: user@example.com
+              - column:
+                  name: password
+                  value: $2a$10$zFoOVRbOX39EDGXC0endRuqO9hvXIaW/EuP0ZCH0waHL7bIgBZ8uu
```
Comment (okuzan) [inline src/main/resources/db/changelog/changes/07-insert-users.yaml:17]: add password in plain text as comment next to this value

File: src/main/java/online/store/book/config/SecurityConfig.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    @Bean
+    public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
+        return http
+                .cors(AbstractHttpConfigurer::disable)
+                .csrf(AbstractHttpConfigurer::disable)
+                .authorizeHttpRequests(
+                        auth -> auth
+                                .requestMatchers("/api/auth/register", "/error",
```
Comment (okuzan) [inline src/main/java/online/store/book/config/SecurityConfig.java]:
```suggestion
                                .requestMatchers("/auth/**", "/error",
```
don't hardcode context prefix + login should be excluded too

**Review by okuzan — CHANGES_REQUESTED** [review]

File: src/main/java/online/store/book/model/Role.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(nullable = false, unique = true)
+    @Enumerated(EnumType.STRING)
+    private RoleName name;
+
+    @Override
+    public String getAuthority() {
+        return name.toString();
```
Comment (Elena-Bruyako) [inline src/main/java/online/store/book/model/Role.java]:
```suggestion
        return name.name();
```

File: src/main/java/online/store/book/model/User.java [hunk 4]
```diff
@@ -26,4 +38,48 @@ public class User {
     @Column(nullable = false)
     private String lastName;
     private String shippingAddress;
+    @ManyToMany(fetch = FetchType.EAGER)
```
Comment (Elena-Bruyako) [inline src/main/java/online/store/book/model/User.java]:
```suggestion
    @ManyToMany
```

File: src/main/java/online/store/book/model/User.java [hunk 5]
```diff
@@ -26,4 +38,48 @@ public class User {
...
+    @JoinTable(
+            name = "users_roles",
+            joinColumns = @JoinColumn(name = "user_id"),
+            inverseJoinColumns = @JoinColumn(name = "role_id")
+    )
+    private Set<Role> roles;
```
Comment (Elena-Bruyako) [inline src/main/java/online/store/book/model/User.java]:
```suggestion
    private Set<Role> roles = new HashSet<>();
```

File: src/main/java/online/store/book/model/User.java [hunk 6]
```diff
@@ -26,4 +38,48 @@ public class User {
...
+    }
+
+    @Override
+    public boolean isCredentialsNonExpired() {
+        return true;
+    }
+
+    @Override
+    public boolean isEnabled() {
+        return true;
```
Comment (Elena-Bruyako) [inline src/main/java/online/store/book/model/User.java]:
```suggestion
        return !isDeleted;
```

File: src/main/java/online/store/book/security/CustomUserDetailsService.java [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+@Service
+@RequiredArgsConstructor
+public class CustomUserDetailsService implements UserDetailsService {
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
+        return userRepository.findUserByEmail(email).orElseThrow(
+                () -> new UsernameNotFoundException("Can't find user by email: " + email));
```
Comment (Elena-Bruyako) [inline src/main/java/online/store/book/security/CustomUserDetailsService.java]:
```suggestion
                () -> new EntityNotFoundException("Can't find user by email: " + email));
```

File: src/main/java/online/store/book/security/CustomUserDetailsService.java [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class CustomUserDetailsService implements UserDetailsService {
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
```
Comment (Elena-Bruyako) [inline src/main/java/online/store/book/security/CustomUserDetailsService.java]:
```suggestion
    public UserDetails loadUserByUsername(String email) {
```

File: src/main/java/online/store/book/security/CustomUserDetailsService.java [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+        return userRepository.findUserByEmail(email).orElseThrow(
```
Comment (Elena-Bruyako) [inline src/main/java/online/store/book/security/CustomUserDetailsService.java:17]: Add to UserRepository method with annotation

```
@EntityGraph(attributePaths = "roles")
Optional<User> findByEmail(String email);
```

File: src/main/java/online/store/book/exceptions/CustomGlobalExceptionHandler.java [hunk 10]
```diff
@@ -87,4 +88,15 @@ public ResponseEntity<Object> handleRegistrationException(
         body.put("errors", ex.getMessage());
         return new ResponseEntity<>(body, HttpStatus.CONFLICT);
     }
+
+    @ExceptionHandler(UsernameNotFoundException.class)
```
Comment (Elena-Bruyako) [inline src/main/java/online/store/book/exceptions/CustomGlobalExceptionHandler.java]:
```suggestion
    @ExceptionHandler(EntityNotFoundException.class)
```

File: src/main/java/online/store/book/exceptions/CustomGlobalExceptionHandler.java [hunk 11]
```diff
@@ -87,4 +88,15 @@ public ResponseEntity<Object> handleRegistrationException(
...
+    public ResponseEntity<Object> handleUsernameNotFoundException(
```
Comment (Elena-Bruyako) [inline src/main/java/online/store/book/exceptions/CustomGlobalExceptionHandler.java]: rename

File: src/main/java/online/store/book/controller/BookController.java [hunk 12]
```diff
@@ -32,12 +33,14 @@ public class BookController {
 
     @Operation(summary = "Get all books", description = "Get all books")
     @GetMapping
+    @PreAuthorize("hasRole('ROLE_USER')")
```
Comment (Elena-Bruyako) [inline src/main/java/online/store/book/controller/BookController.java:36]: You should also implement other controller with POST: /api/auth/register 
for more details check task description

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/online/store/book/controller/BookController.java (hunk 12 above)
Comment (vfilimo) [reply src/main/java/online/store/book/controller/BookController.java:36]: discussed in chat

**Review by Elena-Bruyako — APPROVED** [review]
well done

**Pull request merged**